In [2]:
%matplotlib qt
from bluesky import RunEngine
from bluesky.plan_stubs import mv
from bluesky.plans import scan, list_scan, grid_scan, list_grid_scan, \
    spiral, spiral_square, rel_spiral
from bluesky.callbacks.best_effort import BestEffortCallback
from ophyd.sim import det, det4, motor, motor1, motor2
from matplotlib.pyplot import ion
from bluesky.utils import install_nb_kicker
install_nb_kicker()

ion()

In [3]:
RE = RunEngine()

bec = BestEffortCallback()
RE.subscribe(bec)

0

## Scan

In [3]:
RE(
    scan([det4], motor1, -1.5, 2, motor2, 1.3, -0.75, num=20)
)



Transient Scan ID: 1     Time: 2024-08-28 09:59:57
Persistent Unique Scan ID: '66881be4-b6a4-4f5f-b454-8e848a25fa15'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 09:59:57.2 |     -1.500 |      1.300 |      0.139 |
|         2 | 09:59:57.4 |     -1.316 |      1.192 |      0.207 |
|         3 | 09:59:57.4 |     -1.132 |      1.084 |      0.293 |
|         4 | 09:59:57.4 |     -0.947 |      0.976 |      0.396 |
|         5 | 09:59:57.4 |     -0.763 |      0.868 |      0.513 |
|         6 | 09:59:57.4 |     -0.579 |      0.761 |      0.633 |
|         7 | 09:59:57.4 |     -0.395 |      0.653 |      0.748 |
|         8 | 09:59:57.4 |     -0.211 |      0.545 |      0.843 |
|         9 | 09:59:57.4 |     -0.026 |      0.437 |      0.909 |
|        10 | 09:59:57.4 |      0.158 |      0.329 |      0.936 |
|

('66881be4-b6a4-4f5f-b454-8e848a25fa15',)

## List Scan

In [4]:
RE(
    list_scan([det], motor, [-1.3, -1.2, 0, 1.8, 2], motor1, [1.3, 1.2, -0.2, -1.3, -1.4])
)



Transient Scan ID: 2     Time: 2024-08-28 10:00:03
Persistent Unique Scan ID: 'e7513af7-ab20-45ea-ab33-a62427a30e7a'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |      motor |     motor1 |        det |
+-----------+------------+------------+------------+------------+
|         1 | 10:00:03.9 |     -1.300 |      1.300 |      0.430 |
|         2 | 10:00:03.9 |     -1.200 |      1.200 |      0.487 |
|         3 | 10:00:04.0 |      0.000 |     -0.200 |      1.000 |
|         4 | 10:00:04.0 |      1.800 |     -1.300 |      0.198 |
|         5 | 10:00:04.0 |      2.000 |     -1.400 |      0.135 |
+-----------+------------+------------+------------+------------+
generator list_scan ['e7513af7'] (scan num: 2)





('e7513af7-ab20-45ea-ab33-a62427a30e7a',)

## Grid Scan

In [5]:
RE(
    grid_scan([det4], motor1, -1, 1, 10, motor2, -1, 1, 10)
)



Transient Scan ID: 3     Time: 2024-08-28 10:00:14
Persistent Unique Scan ID: 'e2ca9689-5d6f-4a79-88b0-331d8d8c71fd'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 10:00:14.7 |     -1.000 |     -1.000 |      0.368 |
|         2 | 10:00:14.8 |     -1.000 |     -0.778 |      0.448 |
|         3 | 10:00:14.8 |     -1.000 |     -0.556 |      0.520 |
|         4 | 10:00:14.8 |     -1.000 |     -0.333 |      0.574 |
|         5 | 10:00:14.9 |     -1.000 |     -0.111 |      0.603 |
|         6 | 10:00:14.9 |     -1.000 |      0.111 |      0.603 |
|         7 | 10:00:14.9 |     -1.000 |      0.333 |      0.574 |
|         8 | 10:00:14.9 |     -1.000 |      0.556 |      0.520 |
|         9 | 10:00:15.0 |     -1.000 |      0.778 |      0.448 |
|        10 | 10:00:15.0 |     -1.000 |      1.000 |      0.368 |
|

('e2ca9689-5d6f-4a79-88b0-331d8d8c71fd',)

In [6]:
RE(
    grid_scan([det4], motor1, -5, 5, 30, motor2, -5, 5, 30)
)



Transient Scan ID: 4     Time: 2024-08-28 10:00:28
Persistent Unique Scan ID: 'fdc2dbea-6d71-45ec-b33d-ccfd23482b8e'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 10:00:28.0 |     -5.000 |     -5.000 |      0.000 |
|         2 | 10:00:28.1 |     -5.000 |     -4.655 |      0.000 |
|         3 | 10:00:28.2 |     -5.000 |     -4.310 |      0.000 |
|         4 | 10:00:28.2 |     -5.000 |     -3.966 |      0.000 |
|         5 | 10:00:28.2 |     -5.000 |     -3.621 |      0.000 |
|         6 | 10:00:28.2 |     -5.000 |     -3.276 |      0.000 |
|         7 | 10:00:28.2 |     -5.000 |     -2.931 |      0.000 |
|         8 | 10:00:28.3 |     -5.000 |     -2.586 |      0.000 |
|         9 | 10:00:28.3 |     -5.000 |     -2.241 |      0.000 |
|        10 | 10:00:28.3 |     -5.000 |     -1.897 |      0.000 |
|

|       114 | 10:00:30.3 |     -3.966 |      2.931 |      0.000 |
|       115 | 10:00:30.3 |     -3.966 |      3.276 |      0.000 |
|       116 | 10:00:30.3 |     -3.966 |      3.621 |      0.000 |
|       117 | 10:00:30.3 |     -3.966 |      3.966 |      0.000 |
|       118 | 10:00:30.3 |     -3.966 |      4.310 |      0.000 |
|       119 | 10:00:30.3 |     -3.966 |      4.655 |      0.000 |
|       120 | 10:00:30.4 |     -3.966 |      5.000 |      0.000 |
|       121 | 10:00:30.4 |     -3.621 |     -5.000 |      0.000 |
|       122 | 10:00:30.4 |     -3.621 |     -4.655 |      0.000 |
|       123 | 10:00:30.4 |     -3.621 |     -4.310 |      0.000 |
|       124 | 10:00:30.4 |     -3.621 |     -3.966 |      0.000 |
|       125 | 10:00:30.4 |     -3.621 |     -3.621 |      0.000 |
|       126 | 10:00:30.4 |     -3.621 |     -3.276 |      0.000 |
|       127 | 10:00:30.5 |     -3.621 |     -2.931 |      0.000 |
|       128 | 10:00:30.5 |     -3.621 |     -2.586 |      0.000 |
|       12

+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|       350 | 10:00:34.4 |     -1.207 |      1.552 |      0.145 |
|       351 | 10:00:34.4 |     -1.207 |      1.897 |      0.080 |
|       352 | 10:00:34.4 |     -1.207 |      2.241 |      0.039 |
|       353 | 10:00:34.4 |     -1.207 |      2.586 |      0.017 |
|       354 | 10:00:34.5 |     -1.207 |      2.931 |      0.007 |
|       355 | 10:00:34.5 |     -1.207 |      3.276 |      0.002 |
|       356 | 10:00:34.5 |     -1.207 |      3.621 |      0.001 |
|       357 | 10:00:34.5 |     -1.207 |      3.966 |      0.000 |
|       358 | 10:00:34.5 |     -1.207 |      4.310 |      0.000 |
|       359 | 10:00:34.6 |     -1.207 |      4.655 |      0.000 |
|       360 | 10:00:34.6 |     -1.207 |      5.000 |      0.000 |
|       361 | 10:00:34.6 |     -0.862 |     -5.000 |      0.000 |
|       36

|       583 | 10:00:38.9 |      1.552 |     -0.862 |      0.207 |
|       584 | 10:00:38.9 |      1.552 |     -0.517 |      0.262 |
|       585 | 10:00:38.9 |      1.552 |     -0.172 |      0.296 |
|       586 | 10:00:38.9 |      1.552 |      0.172 |      0.296 |
|       587 | 10:00:38.9 |      1.552 |      0.517 |      0.262 |
|       588 | 10:00:39.0 |      1.552 |      0.862 |      0.207 |
|       589 | 10:00:39.0 |      1.552 |      1.207 |      0.145 |
|       590 | 10:00:39.1 |      1.552 |      1.552 |      0.090 |
|       591 | 10:00:39.1 |      1.552 |      1.897 |      0.050 |
|       592 | 10:00:39.1 |      1.552 |      2.241 |      0.024 |
|       593 | 10:00:39.1 |      1.552 |      2.586 |      0.011 |
|       594 | 10:00:39.1 |      1.552 |      2.931 |      0.004 |
|       595 | 10:00:39.2 |      1.552 |      3.276 |      0.001 |
|       596 | 10:00:39.2 |      1.552 |      3.621 |      0.000 |
|       597 | 10:00:39.2 |      1.552 |      3.966 |      0.000 |
|       59

|       816 | 10:00:43.7 |      4.310 |     -3.276 |      0.000 |
|       817 | 10:00:43.8 |      4.310 |     -2.931 |      0.000 |
|       818 | 10:00:43.8 |      4.310 |     -2.586 |      0.000 |
|       819 | 10:00:43.8 |      4.310 |     -2.241 |      0.000 |
|       820 | 10:00:43.8 |      4.310 |     -1.897 |      0.000 |
|       821 | 10:00:43.8 |      4.310 |     -1.552 |      0.000 |
|       822 | 10:00:43.8 |      4.310 |     -1.207 |      0.000 |
|       823 | 10:00:43.9 |      4.310 |     -0.862 |      0.000 |
|       824 | 10:00:43.9 |      4.310 |     -0.517 |      0.000 |
|       825 | 10:00:43.9 |      4.310 |     -0.172 |      0.000 |
|       826 | 10:00:43.9 |      4.310 |      0.172 |      0.000 |
|       827 | 10:00:43.9 |      4.310 |      0.517 |      0.000 |
|       828 | 10:00:43.9 |      4.310 |      0.862 |      0.000 |
|       829 | 10:00:43.9 |      4.310 |      1.207 |      0.000 |
|       830 | 10:00:44.0 |      4.310 |      1.552 |      0.000 |
|       83

('fdc2dbea-6d71-45ec-b33d-ccfd23482b8e',)

## List Grid Scan

In [7]:
RE(
    list_grid_scan([det4], motor1, [-1, -0.5, 0, 0.5, 1, 1.5], motor2, [-1.5, -1, -0.5, 0, 0.5, 1])
)



Transient Scan ID: 5     Time: 2024-08-28 10:00:54
Persistent Unique Scan ID: 'f5c348b1-8d07-472b-a220-a0270848ab39'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 10:00:54.0 |     -1.000 |     -1.500 |      0.197 |
|         2 | 10:00:54.1 |     -1.000 |     -1.000 |      0.368 |
|         3 | 10:00:54.2 |     -1.000 |     -0.500 |      0.535 |
|         4 | 10:00:54.2 |     -1.000 |      0.000 |      0.607 |
|         5 | 10:00:54.2 |     -1.000 |      0.500 |      0.535 |
|         6 | 10:00:54.3 |     -1.000 |      1.000 |      0.368 |
|         7 | 10:00:54.3 |     -0.500 |     -1.500 |      0.287 |
|         8 | 10:00:54.3 |     -0.500 |     -1.000 |      0.535 |
|         9 | 10:00:54.3 |     -0.500 |     -0.500 |      0.779 |
|        10 | 10:00:54.3 |     -0.500 |      0.000 |      0.882 |
|

('f5c348b1-8d07-472b-a220-a0270848ab39',)

## Spiral Scans

In [7]:
RE(
    spiral(
        detectors=[det4],
        x_motor=motor1,
        y_motor=motor2,
        x_start=0,
        y_start=0,
        x_range=10,
        y_range=10,
        dr=5,
        dr_y=5,
        nth=36
    )
)



Transient Scan ID: 4     Time: 2024-08-28 17:02:48
Persistent Unique Scan ID: '2097a201-7726-4aa2-b5a0-a6dbb87785f2'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 17:02:48.5 |      5.000 |      0.000 |      0.000 |
|         2 | 17:02:48.5 |     -2.500 |      4.330 |      0.000 |
|         3 | 17:02:48.6 |     -3.214 |      3.830 |      0.000 |
|         4 | 17:02:48.7 |     -3.830 |      3.214 |      0.000 |
|         5 | 17:02:48.7 |     -4.330 |      2.500 |      0.000 |
|         6 | 17:02:48.7 |     -4.698 |      1.710 |      0.000 |
|         7 | 17:02:48.7 |     -4.924 |      0.868 |      0.000 |
|         8 | 17:02:48.8 |     -5.000 |      0.000 |      0.000 |
|         9 | 17:02:48.8 |      2.500 |     -4.330 |      0.000 |
|        10 | 17:02:48.8 |      3.214 |     -3.830 |      0.000 |
|

('2097a201-7726-4aa2-b5a0-a6dbb87785f2',)

In [9]:
RE(
    spiral(
        detectors=[det4],
        x_motor=motor1,
        y_motor=motor2,
        x_start=0,
        y_start=0,
        x_range=25,
        y_range=25,
        dr=5,
        dr_y=5,
        nth=10
    )
)



Transient Scan ID: 6     Time: 2024-08-28 17:09:03
Persistent Unique Scan ID: '29c55ab4-e78e-47a7-9f3f-1d17a9ffca19'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 17:09:03.1 |      5.000 |      0.000 |      0.000 |
|         2 | 17:09:03.1 |      4.045 |      2.939 |      0.000 |
|         3 | 17:09:03.2 |      1.545 |      4.755 |      0.000 |
|         4 | 17:09:03.2 |     -1.545 |      4.755 |      0.000 |
|         5 | 17:09:03.3 |     -4.045 |      2.939 |      0.000 |
|         6 | 17:09:03.3 |     -5.000 |      0.000 |      0.000 |
|         7 | 17:09:03.3 |     -4.045 |     -2.939 |      0.000 |
|         8 | 17:09:03.3 |     -1.545 |     -4.755 |      0.000 |
|         9 | 17:09:03.4 |      1.545 |     -4.755 |      0.000 |
|        10 | 17:09:03.4 |      4.045 |     -2.939 |      0.000 |
|

('29c55ab4-e78e-47a7-9f3f-1d17a9ffca19',)

## Square Spiral Scan

In [10]:
RE(
    spiral_square(
        detectors=[det4],
        x_motor=motor1,
        y_motor=motor2,
        x_center=-2,
        y_center=-2,
        x_range=10,
        y_range=10,
        x_num=20,
        y_num=20
    )
)



Transient Scan ID: 8     Time: 2024-08-28 10:01:40
Persistent Unique Scan ID: 'ee0f5ab6-4b71-4584-a5e3-65e6592ae547'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor2 |     motor1 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 10:01:40.1 |     -1.737 |     -2.263 |      0.017 |
|         2 | 10:01:40.2 |     -1.737 |     -1.737 |      0.049 |
|         3 | 10:01:40.2 |     -2.263 |     -1.737 |      0.017 |
|         4 | 10:01:40.2 |     -2.263 |     -2.263 |      0.006 |
|         5 | 10:01:40.2 |     -2.263 |     -2.789 |      0.002 |
|         6 | 10:01:40.2 |     -1.737 |     -2.789 |      0.005 |
|         7 | 10:01:40.3 |     -1.211 |     -2.789 |      0.010 |
|         8 | 10:01:40.3 |     -1.211 |     -2.263 |      0.037 |
|         9 | 10:01:40.3 |     -1.211 |     -1.737 |      0.106 |
|        10 | 10:01:40.3 |     -1.211 |     -1.211 |      0.231 |
|

|       114 | 10:01:43.0 |      0.895 |     -3.316 |      0.003 |
|       115 | 10:01:43.0 |      0.895 |     -2.789 |      0.014 |
|       116 | 10:01:43.0 |      0.895 |     -2.263 |      0.052 |
|       117 | 10:01:43.0 |      0.895 |     -1.737 |      0.148 |
|       118 | 10:01:43.1 |      0.895 |     -1.211 |      0.322 |
|       119 | 10:01:43.1 |      0.895 |     -0.684 |      0.530 |
|       120 | 10:01:43.1 |      0.895 |     -0.158 |      0.662 |
|       121 | 10:01:43.2 |      0.895 |      0.368 |      0.626 |
|       122 | 10:01:43.2 |      0.895 |      0.895 |      0.449 |
|       123 | 10:01:43.2 |      0.368 |      0.895 |      0.626 |
|       124 | 10:01:43.2 |     -0.158 |      0.895 |      0.662 |
|       125 | 10:01:43.2 |     -0.684 |      0.895 |      0.530 |
|       126 | 10:01:43.3 |     -1.211 |      0.895 |      0.322 |
|       127 | 10:01:43.3 |     -1.737 |      0.895 |      0.148 |
|       128 | 10:01:43.3 |     -2.263 |      0.895 |      0.052 |
|       12

+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor2 |     motor1 |       det4 |
+-----------+------------+------------+------------+------------+
|       350 | 10:01:48.7 |      3.000 |     -3.316 |      0.000 |
|       351 | 10:01:48.7 |      3.000 |     -2.789 |      0.000 |
|       352 | 10:01:48.7 |      3.000 |     -2.263 |      0.001 |
|       353 | 10:01:48.8 |      3.000 |     -1.737 |      0.002 |
|       354 | 10:01:48.8 |      3.000 |     -1.211 |      0.005 |
|       355 | 10:01:48.8 |      3.000 |     -0.684 |      0.009 |
|       356 | 10:01:48.9 |      3.000 |     -0.158 |      0.011 |
|       357 | 10:01:48.9 |      3.000 |      0.368 |      0.010 |
|       358 | 10:01:48.9 |      3.000 |      0.895 |      0.007 |
|       359 | 10:01:48.9 |      3.000 |      1.421 |      0.004 |
|       360 | 10:01:49.0 |      3.000 |      1.947 |      0.002 |
|       361 | 10:01:49.0 |      3.000 |      2.474 |      0.001 |
|       36

('ee0f5ab6-4b71-4584-a5e3-65e6592ae547',)

## Relative Spiral

In [8]:
RE(mv(motor1, 3))
RE(mv(motor2, -1))
RE(
    rel_spiral(
        detectors=[det4],
        x_motor=motor1,
        y_motor=motor2,
        x_range=10,
        y_range=10,
        dr=1.5,
        dr_y=1.5,
        nth=10
    )
)



Transient Scan ID: 5     Time: 2024-08-28 17:08:37
Persistent Unique Scan ID: 'a5a77e03-24a1-40f4-a3fc-9d99ace52746'
New stream: 'primary'
+-----------+------------+------------+------------+------------+
|   seq_num |       time |     motor1 |     motor2 |       det4 |
+-----------+------------+------------+------------+------------+
|         1 | 17:08:37.5 |      4.500 |     -1.000 |      0.000 |
|         2 | 17:08:37.5 |      4.214 |     -0.118 |      0.000 |
|         3 | 17:08:37.6 |      3.464 |      0.427 |      0.002 |
|         4 | 17:08:37.6 |      2.536 |      0.427 |      0.037 |
|         5 | 17:08:37.6 |      1.786 |     -0.118 |      0.201 |
|         6 | 17:08:37.6 |      1.500 |     -1.000 |      0.197 |
|         7 | 17:08:37.6 |      1.786 |     -1.882 |      0.035 |
|         8 | 17:08:37.7 |      2.536 |     -2.427 |      0.002 |
|         9 | 17:08:37.7 |      3.464 |     -2.427 |      0.000 |
|        10 | 17:08:37.7 |      4.214 |     -1.882 |      0.000 |
|

('a5a77e03-24a1-40f4-a3fc-9d99ace52746',)

### Exercícios

1. Obtenha um heatmap dos dados do detector 'det5' no intervalo	-4 mm até 8 mm no motor 'jittery_motor1' e no intervalo -6mm até 13mm no motor 'jittery_motor2'. 	

2. Faça 15 aquisições com o detector 'det5' com ponto inicial -2 mm no motor 'jittery_motor1' e -7 mm no motor 'jittery_motor2' e com ponto final 3.5 mm no motor 'jittery_motor1' e 5.2 mm no motor 'jittery_motor2'.  

3. Faça três scans circulares com raio 4, 8 e 12 milimetros com centro -3 mm no motor 'jittery_motor1' e -2 mm no motor 'jittery_motor2', faça cada aquisição com 10º graus de difereça entre os pontos. 

4. 